src/esp32/hcsr04.rs

```diff
@@ -0,0 +1,289 @@
+                    "HCSR04Sensor: error handling `timeout_ms` value: {:?}",
+                    e
+                )),
+            },
+            |v| Ok(Some(Duration::from_millis(v as u64))),
+        )?;
+
+        Ok(Arc::new(Mutex::new(HCSR04Sensor::new(
+            board,
+            trigger_pin,
+            echo_interrupt_pin,
+            timeout,
+        )?)))
+    }
+
+    fn new(
+        board: BoardType,
+        trigger_pin: i32,
+        echo_interrupt_pin: i32,
+        timeout: Option<Duration>,
+    ) -> anyhow::Result<HCSR04Sensor> {
+        // TODO: Move this into micro-rdk and unify with pin.rs
+        init_isr_alloc_flags(esp_idf_hal::interrupt::InterruptType::Iram.into());
+        enable_isr_service()?;
+
+        let notification = Notification::new();
+        let notifier = notification.notifier();
+
+        let mut sensor = Self {
+            _board: board,
+            trigger_pin: PinDriver::output(unsafe { AnyIOPin::new(trigger_pin) })?,
+            echo_interrupt_pin: PinDriver::input(unsafe { AnyIOPin::new(echo_interrupt_pin) })?,
+            timeout: timeout.unwrap_or(Duration::from_millis(50)),
+            interrupt_notification: notification,
+            isr_shared_state: Arc::new(IsrSharedState {
+                timestamp: 0.into(),
+                notifier,
+            }),
+        };
+
+        sensor.echo_interrupt_pin.set_pull(Pull::Down)?;
+
+        // Start the trigger pin high: the pulse is sent on the
+        // falling edge, so we can just go low immediately in
+        // `get_readings` to send it. As long as we don't get another
+        // `get_readings` request within 10us of the prior one
+        // completing, the pin will be high long enough to trigger the
+        // pulse.
+        sensor.trigger_pin.set_high()?;
+
+        sensor
+            .echo_interrupt_pin
+            .set_interrupt_type(InterruptType::AnyEdge)?;
+
+        unsafe {
+            esp!(gpio_isr_handler_add(
+                echo_interrupt_pin,
+                Some(Self::subscription_interrupt),
+                Arc::as_ptr(&sensor.isr_shared_state) as *mut _,
+            ))?;
+        }
+
+        Ok(sensor)
+    }
+
+    #[inline(always)]
+    #[link_section = ".iram1.intr_srv"]
+    unsafe extern "C" fn subscription_interrupt(arg: *mut core::ffi::c_void) {
+        let arg: &mut IsrSharedState = &mut *(arg as *mut _);
+        let when = esp_idf_sys::esp_timer_get_time();
+        match arg
+            .timestamp
+            .compare_exchange(0, when, Ordering::AcqRel, Ordering::Acquire)
+        {
+            Ok(_) => {
+                // Initial edge: timestamp gets stored.
+            }
+            Err(-1) => {
+                // Locked out: ignore this interrupt. Sometimes it
+                // seems that there are spurious edges on timeouts.
+            }
+            Err(prior) => {
+                // Final edge: try set the lockout to debounce
```

Comment:
Typo here, but I'm hoping that the 'lockout' state goes away so maybe this comment does too.

Reply:
I removed the lockout state, since after re-testing with it removed I no longer find it to be necessary, so this code is gone.